feat: add blob-stream queue type with remote storage backend

viktorerlingsson opened the pull request:
Adds a new `blob-stream` queue type (`x-queue-type: "blob-stream"`) that offloads sealed segments to S3-compatible object storage, enabling virtually unlimited retention with bounded local disk usage. Regular `stream` queues are unaffected — both types can coexist in the same instance.

Active (unsealed) segments stay on local disk; once sealed they are uploaded to remote storage and may be evicted locally based on cache size.

Configuration is per-deployment via `[blob-storage]` in `lavinmq.ini` (or matching `--blob-storage-*` CLI flags).

Implements #1070

## Components

- `BlobStorageClient` — minimal S3-compatible client (PUT, GET, multi-object DELETE, LIST) with upload retry.
- `BlobSegmentCache` — local cache of segments fetched from remote storage, size-bounded eviction, read-ahead prefetch for sequential consumers.
- `BlobMessageStore` — extends the stream message store to resolve offsets across remote-only segments, maintains an append-only segment catalog for fast startup (avoids a full bucket LIST), and coordinates async uploads of sealed segments.
- `BlobStream` — new queue class inheriting from `Stream`, always uses `BlobMessageStore`.

## Configuration

```ini
[blob-storage]
region = eu-north-1
endpoint = https://my-bucket.s3.eu-north-1.amazonaws.com
access_key_id = AKIAIOSFODNN7EXAMPLE
secret_access_key = wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
local_segments_per_stream = 50
```

Endpoint must be virtual-hosted-style (bucket in hostname). All keys also available as `--blob-storage-*` CLI flags.

## Testing

Specs cover upload, eviction, offset resolution across remote-only segments, restart/catalog recovery, and overflow edge cases. A test server under `spec/support/` backs the specs — no live endpoint required.

Additionally, a manual integration test suite under `extras/s3-integration/` exercises the feature end-to-end against a real S3-compatible bucket (needs re-run after refactor).

## Dependencies

- Adds `awscr-signer` for AWS SigV4 signing.

## Open items

- [ ] Tested against AWS S3 only; other providers (MinIO, R2, GCS, B2) not yet verified.
- [ ] Low consume throughput with very small segments (<1MB) — see test report for details.
- [x] No upload backpressure: sustained remote outage grows upload queue unboundedly.
- [ ] Consumer offsets are not persisted to remote storage.

lukas8219 reviewed (commented):
just small comments since i haven't been able to ran it locally yet

spuun reviewed (commented):
Dunno if it would be better or not, but maybe it's good to have global download_segments and monitor_downloads fibers. Not for a first version of s3 queue thought.

viktorerlingsson committed:
feat: add S3 storage backend for streams

Streams can optionally offload sealed segments to S3-compatible object
storage, enabling virtually unlimited retention with bounded local disk
usage. Active (unsealed) segments remain on local disk; once sealed,
they are uploaded to S3 and may be evicted locally based on cache size.

Configuration is per-deployment via `[streams]` options in lavinmq.ini
(endpoint, bucket, region, credentials, cache size, prefetch). When
unset, streams behave as before with purely local storage.

Components:

* `S3StorageClient` — minimal S3 client covering PUT, GET (with ranges),
  multi-object DELETE, and LIST, with upload retry.
* `S3SegmentCache` — local cache of segments fetched from S3 with
  size-bounded eviction and read-ahead prefetch for sequential consumers.
* `S3MessageStore` — extends the stream message store to resolve offsets
  across S3-only segments, maintains an append-only segment catalog for
  fast startup (avoiding a full bucket LIST), and coordinates async
  uploads of newly sealed segments.

Also fixes a crash in `open_new_segment`: `drop_overflow` was called
before `@segment_first_offset` was set for the new segment. With a tight
`x-max-length`, all prior segments could be dropped, leaving
`@segment_first_offset` empty when `cleanup_consumer_offsets` called
`first_value`. The new segment's metadata is now recorded before
overflow handling runs.

Includes specs covering upload, eviction, offset resolution across
S3-only segments, restart/catalog recovery, and overflow edge cases.
A test S3 server under `spec/support/` backs the specs without requiring
a live S3 endpoint.

viktorerlingsson commented:
## S3 integration test report

All 15 end-to-end tests pass as of 2026-04-13 (harness under `extras/s3-integration/`, run against a real AWS S3 bucket in `eu-north-1`).

### Small segments (512KB, 50K messages)

| # | Test | Result |
|---|------|--------|
| 01 | Publish & async S3 upload | PASS — 101 segments uploaded, ~50K msgs/s publish |
| 02 | Restart & durability (no local segments) | PASS — 49995/50000 recovered, ~7.1K msgs/s consume |
| 03 | Consume purely from S3 | PASS — 49995 consumed at ~8–13K msgs/s, local cap respected |
| 04 | 4-consumer benchmark from S3 | PASS — ~800–1700 msgs/s per consumer under contention |
| 05 | Mixed workload (consume old + publish new) | PASS — 69995 consumed, 40 new segments uploaded |
| 06 | Segment catalog upload + recovery | PASS — catalog uploaded on shutdown, restored on restart |

### Large segments (8MB, 1.7M messages)

| # | Test | Result |
|---|------|--------|
| 10 | Large publish & S3 upload | PASS — 214 segments uploaded, ~63K msgs/s publish |
| 11 | Restart with partial local data | PASS — 1.7M survive, S3 fills gaps |
| 12 | Restart with no local data | PASS — 1.7M recovered from S3 alone |
| 13 | Full consume x3 from S3 | PASS — steady 80–130K msgs/s (peaks 370K when cached) |

### Recovery & edge cases

| # | Test | Result |
|---|------|--------|
| 07 | SIGKILL + catalog deletion | PASS — recovers from individual `.meta` files in S3 |
| 08 | Stream deletion (local + S3 cleanup) | PASS — 203 S3 objects deleted via multi-object DELETE, no segfault |
| 09 | `x-max-length` eviction during upload | PASS — regression test for the `open_new_segment` crash fix |
| 14 | 4 concurrent S3-backed streams | PASS — all 4 recover 1.7M each after no-local-data restart |
| 15 | Offset resolution from S3-only segments | PASS — integer, `"first"`, `"last"` all resolve |

Timestamp-based offsets are not exercised by the harness (`lavinmqperf --consumer-args` is JSON-only, no timestamp type). The code path goes through `find_offset_in_segments` just like integer offsets, so integer coverage is representative.

### Known issues

**Low consume throughput with small segments.** Tests 1–5 show only ~8–13K msgs/s consume from S3 with 512KB segments, vs. 80–130K msgs/s with 8MB segments. Each 512KB segment holds ~500 messages so S3 round-trip overhead dominates. Manual testing hits 50–60K msgs/s in similar conditions, so there's likely headroom — prefetch budget / download concurrency / cache coordination worth investigating.

### Architecture concerns worth flagging

1. **Init complexity.** `S3MessageStore#initialize` is a multi-phase sequence (list S3, load catalog, download meta, download last segment, `super`, load stats, fix offsets, start workers, upload remaining, create cache). Ordering is subtle; partial failures leave unclear state. Candidate for an explicit state-machine refactor.
2. **Concurrent mutation of shared hashes.** `@s3_segments`, `@segment_msg_count`, `@segment_first_offset` are touched by upload workers, download workers, the cache, and the main fiber. Only the cache's own dict uses a `Mutex`; the rest relies on cooperative scheduling. Any S3 HTTP call yields — latent race. Not observed in tests but worth auditing.
3. **Inheritance coupling.** `S3MessageStore` subclasses `StreamMessageStore` and overrides ~10 methods. Tight coupling to base-class internals; refactors of `StreamMessageStore` can silently break S3 behavior. Composition would isolate better.
4. **No upload backpressure.** If rotation outpaces upload throughput (slow network, S3 outage), `@upload_queue` and `@failed_uploads` grow unboundedly. Needs a bounded channel or eviction cap.

viktorerlingsson committed:
refactor: simplify S3 segment fetch by removing wait_for_segment

Shared @download_complete channel had a signal-stealing race: a waiter
for segment A could consume the completion notification for segment B,
causing B's waiter to time out. Fix by removing the wait path entirely.

Consumer-initiated downloads now go straight to a direct fetch. If a
cache prefetch is racing for the same segment, the filesystem rename in
S3StorageClient#download_segment arbitrates — the loser's bytes are
discarded. Wasted GET in a narrow window, not a correctness bug.

viktorerlingsson committed:
refactor: introduce blob-stream queue type and rename S3 references

Break S3-backed streams out into a separate `blob-stream` queue type
(`x-queue-type: "blob-stream"`) so regular streams and blob streams can
coexist in the same instance. Rename S3-specific naming to
storage-agnostic "blob" terminology throughout.

- Add `BlobStream < Stream` class that always uses `BlobMessageStore`
- Register `blob-stream` queue type in `QueueFactory`
- Remove S3 conditional from `Stream#init_msg_store`
- Rename classes: S3MessageStore → BlobMessageStore,
  S3StorageClient → BlobStorageClient, S3SegmentCache → BlobSegmentCache
- Rename config section `[s3-storage]` → `[blob-storage]` and all
  associated CLI flags and properties
- Remove global `streams_s3_storage` enable flag (queue type is explicit)
- Rename internal identifiers from s3 to remote/blob terminology

viktorerlingsson committed:
chore: add blob-stream routing to queue list and detail UI pages

viktorerlingsson committed:
refactor: parallelize blob upload workers and harden retry logic

Move upload workers onto a dedicated Parallel execution context so HTTP
uploads run on a shared thread pool instead of fibers on the default
scheduler. Bound the upload queue to 64 to apply backpressure.

Replace the rotation-coupled @failed_uploads deque with a timed retry:
after 3 failed attempts, a fiber re-enqueues the segment after a 5s
backoff. Guard @remote_segments and the catalog append with a mutex now
that multiple threads can complete uploads concurrently.
